text/0008-scoped-gems.md

```diff
@@ -0,0 +1,229 @@
+# RubyGems Scoped Gems and Organizations
+
+* Feature Name: Scoped Gems
+* Start Date: 4/1/2022
+* RFC PR:
+* Bundler Issue:
+
+# Summary
+
+ Scoped gems is a feature that allows for grouping related gems together under an organization account. Scoped gems are identified by their name following a specific pattern of `gem@scope`. They can be built, installed, and required the same way as existing gems today. An Organization is a group of users that can own 0 or many gem scopes and 0 or many gems. Only the users that are part of an organization are permitted to publish a new gem under a scope that is registered to that organization (first-time publish).
+
+# Motivation
+
+There are three main motivations for supporting scoped gems and organizations:
+
+1. Reduce customer confusion for tools and services spanning multiple gems. It is not uncommon for an organization to distribute many packages, such as Rails with the ActiveX packages, AWS with SDK clients, Ruby standard library gems, Azure SDKs, etc. Scoped gems are a good way to signal official packages from organizations.
+2. Help reduce [typo-squatting](https://blog.reversinglabs.com/blog/mining-for-malicious-ruby-gems) with malicious code. Scoped gems are also susceptible to typos, however, gems published under a scope are verified to be owned by that organization.
+3. An organization allows for a better gem owner permissions model. Gems owned by the same company or organization currently need to manage users for each gem. Users and gem owners can now be part of a single entity, and that entity can be the owner of the gem.
+
+# Guide-level explanation
+
+## Scoped gems
+
+A scoped gem is any gem that is named following the pattern `gem@scope`, where “gem” is the name of the gem, and “scope” is an organization's reserved gem scope. Gem scopes are globally unique across all organizations.
+
+As an example, consider a scoped gem defined as `hola@mullermp.gemspec`. Note that both the gemspec’s `name` and the required file `lib/hola@mullermp.rb` follows the `gem@scope` pattern:
```

Comment:
Is the lib filename _requirement_ necessary?  I can see it being _recommended_, but making it a requirement makes some use cases more complicated.  For example, in the case of soft forking I'm essentially "replacing" the official gem with my copy, which also implies I don't expect callers to change the thing they are going to require.

Similarly I don't think the gemspec file name _requirement_ is necessary, but could be _recommended_, as I thought the final gem file name is produced from the `spec.name`.

Reply:
I wouldn't call them "requirements" but they are both recommended and by convention, so I would argue that you'd want to do them. The file name doesn't matter even today (you're correct that it uses `spec.name`), and the lib file's name is the same as the gem name as per the "Make your own gem" guide in RubyGems:
```
The convention is to have one Ruby file with the same name as your gem, since that gets loaded when require 'hola' is run.
```

If I understand what you mean about soft forking, let's say you want to replace `hola@mullermp` with your own implementation. You could have both `lib/hola@fryguy.rb` and `lib/hola@mullermp.rb` in your gem where `lib/hola@mullermp.rb` simply does `require 'hola@fryguy'`, which then loads the custom gem code that you've provided.